synapse/rest/client/devices.py

```diff
@@ -283,19 +369,58 @@ async def on_GET(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
         else:
             raise errors.NotFoundError("No dehydrated device available")
 
+    async def on_DELETE(self, request: SynapseRequest) -> Tuple[int, JsonDict]:
+        requester = await self.auth.get_user_by_req(request)
+
+        dehydrated_device = await self.device_handler.get_dehydrated_device(
+            requester.user.to_string()
+        )
+
+        if dehydrated_device is not None:
+            (device_id, device_data) = dehydrated_device
+
+            await self.device_handler.rehydrate_device(
+                requester.user.to_string(),
+                self.auth.get_access_token_from_request(request),
+                device_id,
+            )
+
+            result = {"device_id": device_id}
+
+            return 200, result
+        else:
+            raise errors.NotFoundError("No dehydrated device available")
+
     class PutBody(RequestBodyModel):
         device_data: DehydratedDeviceDataModel
+        device_id: Optional[StrictStr]
```

Comment:
This looks like we are allowing the clients to specify a dehydrated device ID. Sanity check: is there any risk that Alice and Bob can create dehydrated devices with the same ID, and in doing so clobber each other's devices?

From a quick look at this and the existing source, it seems that dehydrated devices are always keyed off a (user, device) pair, so it should be fine... but it might be worth double checking.

Reply:
Devices are keyed off a user, yes, so Alice and Bob clashing can't happen. What can happen is that two Alice's devices try to create a dehydrated device, which is fine, one will overwrite the other and if they use the public key of the device as the device id a clash is very unlikely.